src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,116 @@
 package core.basesyntax;
 
+import java.util.NoSuchElementException;
+
 public class ArrayList<T> implements List<T> {
+    private static final int ZERO = 0;
+    private static final int AUXILIARY_ONE = 1;
+    private static final int DEFAULT_SIZE = 10;
+    private static final double MULTIPLICATION = 1.5;
+    private Object[] elements;
+    private int size;
+
+    public ArrayList() {
+        elements = new Object[DEFAULT_SIZE];
+    }
+
     @Override
     public void add(T value) {
-
+        if (size == elements.length) {
+            int newCapacity = (int) (elements.length * MULTIPLICATION);
+            Object[] newElements = new Object[newCapacity];
+            System.arraycopy(elements, 0, newElements, 0, size);
+            elements = newElements;
+        }
+        elements[size++] = value;
     }
 
     @Override
     public void add(T value, int index) {
-
+        if (index < ZERO || index > size) {
+            throw new ArrayListIndexOutOfBoundsException("Index out of bounds: " + index);
+        }
+        if (size == elements.length) {
+            int newCapacity = (int) (elements.length * MULTIPLICATION);
+            Object[] newElements = new Object[newCapacity];
+            System.arraycopy(elements, 0, newElements, 0, size);
+            elements = newElements;
+        }
+        System.arraycopy(elements, index, elements, index + AUXILIARY_ONE, size - index);
+        elements[index] = value;
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        int newSize = size + list.size();
+        if (newSize > elements.length) {
+            int newCapacity = (int) (elements.length * MULTIPLICATION);
+            if (newCapacity < newSize) {
+                newCapacity = newSize;
+            }
+            Object[] newElements = new Object[newCapacity];
+            System.arraycopy(elements, 0, newElements, 0, size);
+            elements = newElements;
+        }
+        for (int i = ZERO; list.size() > i; i++) {
+            elements[size++] = list.get(i);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        throwException(index);
```

Comment:
What this method does? It checks the index I hope
![image](https://github.com/mate-academy/jv-array-list/assets/20326858/116e2afd-cef8-4244-b0cb-2f546f163d63)